### PR TITLE
KAFKA-3340: add some enhancements to MockConsumer

### PR DESCRIPTION
**Pull request by fhussonnois**
The MockConsumer class should support adding records concurrently.
This allow to implement more complex test scenarios in which records are added concurrently with the records are polled.

In addition, I've implemented the rebalance callback in simple manner.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 1]
```diff
@@ -55,10 +57,12 @@
 
     private AtomicBoolean wakeup;
 
+    private AtomicBoolean rebalanced = new AtomicBoolean(false);
```
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: This is initialized with false, however once set true, it is never set to true again. Is the usage as you intended?

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 2]
```diff
@@ -75,9 +79,14 @@ public MockConsumer(OffsetResetStrategy offsetResetStrategy) {
 
     /** Simulate a rebalance event. */
     public void rebalance(Collection<TopicPartition> newAssignment) {
-        // TODO: Rebalance callbacks
         this.records.clear();
         this.subscriptions.assignFromSubscribed(newAssignment);
+        if( rebalanced.get() ) {
```
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: compareAndSet should probably be used here.

**Comment by SinghAsDev** [comment]
Do you want to create a JIRA first? Are you planning to add some test that actually needs this refactoring?

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java (hunk 1 above)
Comment (fhussonnois) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: This boolean is only used for detecting the first rebalance. In that case we should not invoke the invoke method.

**Comment by fhussonnois** [comment]
Currently, I am using my own class with these modifications to be able to execute multi-threaded tests. In addition, I need to implement some business logic wihtin the rebalance listener which I need to test too. I will create a JIRA for these modifications.

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java (hunk 2 above)
Comment (fhussonnois) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Ok for that change

**Comment by fhussonnois** [comment]
The Jira is created : https://issues.apache.org/jira/browse/KAFKA-3340
The commit messages has been updated

**Comment by tealamore** [comment]
It's been almost 2 years since anything has been done to this PR. It may be safe to just close this.

**Pull request closed without merging**